Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c

```diff
@@ -314,6 +316,220 @@ InternalEfiShellStartCtrlSMonitor(
+
+  @param[out] StartImageStatus  Returned status from gBS->StartImage.
+
+  @retval EFI_SUCCESS           The command executed successfully. The status code
+                                returned by the command is pointed to by StatusCode.
+  @retval EFI_INVALID_PARAMETER The parameters are invalid.
+  @retval EFI_OUT_OF_RESOURCES  Out of resources.
+  @retval EFI_UNSUPPORTED       Nested shell invocations are not allowed.
+**/
+EFI_STATUS
+ShellExecuteMemoryDevicePath(
+  IN CONST EFI_HANDLE               *ParentImageHandle,
+  IN CONST MEMMAP_DEVICE_PATH       *DevicePath,
+  IN CONST CHAR16                   *CommandLine OPTIONAL,
+  IN CONST CHAR16                   **Environment OPTIONAL,
+  OUT EFI_STATUS                    *StartImageStatus OPTIONAL
+  )
+{
+  EFI_STATUS                    Status;
+  EFI_STATUS                    StartStatus;
+  EFI_STATUS                    CleanupStatus;
+  EFI_HANDLE                    NewHandle;
+  EFI_LOADED_IMAGE_PROTOCOL     *LoadedImage;
+  LIST_ENTRY                    OrigEnvs;
+  EFI_SHELL_PARAMETERS_PROTOCOL ShellParamsProtocol;
+  CHAR16                        *ImagePath;
+  UINTN                         Index;
+  CHAR16                        *Walker;
+  CHAR16                        *NewCmdLine;
+
+  if (ParentImageHandle == NULL) {
+    return (EFI_INVALID_PARAMETER);
+  }
+
+  InitializeListHead(&OrigEnvs);
+  ZeroMem(&ShellParamsProtocol, sizeof(EFI_SHELL_PARAMETERS_PROTOCOL));
+
+  NewHandle = NULL;
+
+  NewCmdLine = AllocateCopyPool (StrSize (CommandLine), CommandLine);
+  if (NewCmdLine == NULL) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  for (Walker = NewCmdLine; Walker != NULL && *Walker != CHAR_NULL ; Walker++) {
+    if (*Walker == L'^' && *(Walker+1) == L'#') {
+      CopyMem(Walker, Walker+1, StrSize(Walker) - sizeof(Walker[0]));
+    }
+  }
+
+  //
+  // Load the image with:
+  // FALSE - not from boot manager,
+  // address and size of memory region where the image is located
+  //
+  Status = gBS->LoadImage(
+    FALSE,
+    *ParentImageHandle,
+    (EFI_DEVICE_PATH_PROTOCOL*)DevicePath,
+    (VOID*) ((UINTN)(DevicePath[0].StartingAddress)),
+    DevicePath[0].EndingAddress - DevicePath[0].StartingAddress,
+    &NewHandle);
+
+  if (EFI_ERROR(Status)) {
+    if (NewHandle != NULL) {
+      gBS->UnloadImage(NewHandle);
+    }
+    FreePool (NewCmdLine);
+    return (Status);
+  }
+  Status = gBS->OpenProtocol(
+    NewHandle,
+    &gEfiLoadedImageProtocolGuid,
+    (VOID**)&LoadedImage,
+    gImageHandle,
+    NULL,
+    EFI_OPEN_PROTOCOL_GET_PROTOCOL);
+
+  if (!EFI_ERROR(Status)) {
+    //
+    // If the image is not an app abort it.
+    //
+    if (LoadedImage->ImageCodeType != EfiLoaderCode){
+      ShellPrintHiiEx(
+        -1,
+        -1,
+        NULL,
+        STRING_TOKEN (STR_SHELL_IMAGE_NOT_APP),
+        ShellInfoObject.HiiHandle
+      );
+      goto UnloadImage;
+    }
+
+    ASSERT(LoadedImage->LoadOptionsSize == 0);
+    if (NewCmdLine != NULL) {
```

Comment:
@krystian-hebel this check seems to be redundant? You already check if NewCmdLine is NULL right after getting this memory. I don't see why you do it it again.